Towards a canonical definition of BDD (WIP)

mattwynne opened the pull request:
In my shiny new job as a BDD Advocate at SmartBear, I have plenty of time to focus on helping the world learn the delights of BDD.

It's always bothered me that there was no single place you could point someone to help them understand what BDD is. I'm aiming to build that here on Cucumber's docs.

Together with the help of @sebrose, I've done my best to write a brief summary of the key ideas as we tend to teach them these days. My plan is to build out from here and write more pages that go into more depth, but I'd like some feedback on this introductory page first. Please ignore the links at the bottom to sub-pages, I'm ignoring those for now.

I'd like to make this into a page that all of us in the community are happy with, and I will be interested to hear differences of opinion - as always, the places we see things differently as the places we can learn 😁 

You can see a preview [here](https://deploy-preview-376--cucumber.netlify.com/docs/bdd/).

mattwynne committed:
WIP draft of new BDD content

mattwynne committed:
Add alt attribute for image

mattwynne committed:
Work on BDD index page

mattwynne committed:
Remove old overview that's been replaced by the index page

(and the myths page)

mattwynne committed:
Fix html proofing

mattwynne committed:
Fix another broken link

mattwynne committed:
Fix html

mattwynne committed:
Better 3 practices diagram

mattwynne commented:
@tastapod I would love your thoughts on this.

mattwynne commented:
@richardlawrence I would love your feedback on this too.

mattwynne commented:
@ciaranmcnulty any thoughts on this?

mattwynne commented:
@gojko I'd love your feedback on this.

gojko commented:
after a brief glance (I'll send more info soon), I think you're missing something around building a shared understanding across roles, and something around speed and how it fits into short iterations. Playing the devil's advocate here (I know this is not what you intended, but put on a newbie hat and forget the context), just read the three bullet points on the intro section of that page:

```
- Making a deliberate effort to avoid accidental discovery of business requirements late in the process.
- Eliminating defects and rework to achieve a confident, sustainable pace.
- Producing system documentation that is current, accurate and easy to understand.
```

All of this holds true as the intent of waterfall analysis and producing voluminous documentation upfront. I think the intro should clearly talk about what's unique about BDD here, which is collaboration across roles and speed.

I think you're also missing a major part of what comes after automation. in the book I called those two sections validating frequently and evolving a documentation system - so kind of plugging it into a frequent feedback loop and then refactoring both the docs and the automation as knowledge evolves. that is a major difference between this approach and letting the testing specialist team die with their obsolete tests in many large orgs.

lunivore commented:
I prefer Gojko's "Specification" to the word "Formulation", which just means shaping or creating a stricture; that's a thing that happens all the way through.

mattwynne commented:
@gojko thanks. I totally agree about the speed thing especially. I find people often seem to miss this idea about small pieces and flow, perhaps because those of us with the curse of knowledge take it for granted so much we forget to point it out. So I'll see about bumping that point up from the "principles" into that main intro pitch. Same with the collaboration stuff.

Point taken about the ongoing "habitation" of the symbiotic code-and-living-documentation system, I'm going to play with how to communicate that.

@lunivore thanks, I think I agree with you about "Specification" vs "Formulation". What do you think about "Automation"?

mattwynne committed:
Responding to Gojko and Liz's feedback

I've added the practice of "habitation" and renamed "formulation" to
"specification".

mattwynne committed:
Tweaks

mattwynne committed:
More tweaks

mattwynne committed:
Tweaks

lunivore commented:
I'm OK with "Automation" though I tend to use "Test" as the third thing, since I'm doing BDD with non-software teams too (non-automated examples can also be used to test things!) But worth mentioning automation there even if it isn't in the title.

Also for the first one, since discovery is what happens but not what you actually do, may I suggest "Exploration"? That's the act of deliberately discovering things (as opposed to accidentally doing so).

mattwynne committed:
Add habitation diagram

mattwynne commented:
@gojko I may have mis-represented what you meant by "validating frequently" and "evolving a documentation system" (I don't have your book to hand today) but I've tried to represent them as a fourth, emergent practice of "Habitatition". Your thoughts appreciated.

mattwynne commented:
re "Exploration" @lunivore let's try it!

mattwynne committed:
Try renaming "Discovery" to "Exploration" at Liz's suggestion

mattwynne commented:
@lunivore re "Automation" or "Test" my worry is that neither of those words conveys to (software) people that TDD is what we intend for them to do here. Does that concern you?

mattwynne committed:
Trigger another Netlify build. Preview's not working...

mattwynne commented:
Odd. The Netlify preview seems to have stopped working properly. The preview of the latest commit is [here](https://5d71570a761226000888265d--cucumber.netlify.com/docs/bdd/).

mattwynne commented:
@tomstuart 👋

mattwynne committed:
Calling them specification workshops is confusing now

gasparnagy commented:
I am generally fine with the structure, but I will read it through more in detail. 

Finding the right keywords is always hard. We were thinking about this quite much with @sebrose when we called the three steps as discovery, formulation and automation in our book and we have used these terms pretty successfully to teach many people since them. As I have mentioned already once when we wanted to rename "scenario" in Gherkin, I think the point is not which English word will exactly describe what we do, because there will be no single English word that can describe such a sophisticated activity, but whether we can define the context where those words get their clear meaning. The discovery, formulation, automation trio has already some sort of community acceptance (I hear people talking about BDD using these terms), so we should consider whether the arguments of renaming are strong enough to destroy this. 

From this perspective, for me "Exploration" is not much more precise then "Discovery", just to see the same thing from different angles. In our book we described this as "Explore and discover details through examples". 

On the "specification" vs. "formulation", we were trying to use a word that is not that much overused. Specification has a lot of attached meaning and typically totally different than what we do there. It would be very confusing to teach people about the "BDD specifiaction" which is not the kind of "specification" that we would like to drive you away from. We were also considering "documentation" for this activity, but dropped in favor of formulation because of the same reason. Formulation means by Merrian-Webster: "to put into a systematized statement or expression", that we have found pretty much describing what we do there.

tomstuart commented:
👋 

Thank you for writing this, Matt! It’s really helpful.

You won’t be surprised to hear that I’m not very interested in the correct choice of Billable Nouns, so naturally I don’t believe you should be either — it’s best to [use the words normal people would use](https://gilest.org/normal-words.html). For example: if someone asked you in the pub what “exploration” actually means, what would you say? Does it mean “finding out what everyone wants”, for example? You could just write that instead.

I also think it’s sensible to focus on software (i.e. automated testing) for the sake of clarity. You shouldn’t be coy about Gherkin being the “medium that can be read by both humans and computers”, because that’s useful and actionable information. I think “document those examples in a way that can be automated” is confusing for people who don’t already understand what you’re describing; what is it about the examples that will be automated?

I really like the “four practices” section because it gets very close to answering the single most important question, which is “what should I _actually do_ if I want to try BDD?”. There isn’t yet quite enough there to straightforwardly answer that, but you say you’re going to get into more concrete detail later, so as an overview it works well.

Overall, a great step forward! 👏

mattwynne commented:
Thanks for your input @gasparnagy and @tomstuart.

Tom, while I take your point about the ROI of fiddling around with words, what's important to me here is that this is a definition we can all live with and would want to point people to. So I'm more bothered about consensus and buy-in than wordsmithing. I'm definitely going to apply the "words normal people would use" filter in my next iteration - I love the idea thanks 👍 

Back to those labels, I would like to try and settle on one-word names for the four practices (if we can agree on the four conceptually!) @gasparnagy it sounds to me like you have put a lot of thought into this already, and you make a really good case. I guess it's over to @lunivore - can you live with those terms from Seb and Gaspar's book? Maybe not perfect but good enough?

richardlawrence commented:
I like how this is coming together, and I agree with most of the feedback so far. One big thing jumps out at me in the latest version...

In the "four practices" session, you mention _iterative_ at the top and suggest it in the picture. But then, as I read the descriptions of each from a beginner's perspective, thinking about common misunderstandings I see in my clients, I get the sense that all the examples for a story are identified and specified in a large batch and then implemented one at a time. 

In practice, I see good teams exploring examples a bit, then taking one or two into Gherkin, implementing them, then discovering more examples and exploring those a bit, etc.

It would be great if there were a way to communicate that the actual practice looks more like continuous collaboration than linear steps.

lunivore commented:
I don't think it's necessary to have consensus on the words anyway, and a
bit late for me since I've already gone with Exploration, Specification,
Test :)  https://lizkeogh.com/2015/12/14/bdd-a-three-headed-monster/

FWIW those are also a match for my "Having the conversation (exploration)
is more important than capturing the conversation (specification) is more
important than automating the conversation (test)". I tend to teach both
bits together just to keep the emphasis on the conversation.

On Fri, Sep 6, 2019 at 2:24 PM Matt Wynne <notifications@github.com> wrote:

> Thanks for your input @gasparnagy <https://github.com/gasparnagy> and
> @tomstuart <https://github.com/tomstuart>.
>
> Tom, while I take your point about the ROI of fiddling around with words,
> what's important to me here is that this is a definition we can all live
> with and would want to point people to. So I'm more bothered about
> consensus and buy-in than wordsmithing. I'm definitely going to apply the
> "words normal people would use" filter in my next iteration - I love the
> idea thanks 👍
>
> Back to those labels, I would like to try and settle on one-word names for
> the four practices (if we can agree on the four conceptually!) @gasparnagy
> <https://github.com/gasparnagy> it sounds to me like you have put a lot
> of thought into this already, and you make a really good case. I guess it's
> over to @lunivore <https://github.com/lunivore> - can you live with those
> terms from Seb and Gaspar's book? Maybe not perfect but good enough?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cucumber/docs.cucumber.io/pull/376?email_source=notifications&email_token=AAH5TEFUVYWCV2YZLC7M5RTQIJKZ7A5CNFSM4ITKXLMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6C2R3Q#issuecomment-528853230>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAH5TEAMYF3RV5LY6QLJYH3QIJKZ7ANCNFSM4ITKXLMA>
> .
>


-- 
Liz Keogh
liz@lunivore.com
http://lizkeogh.com
http://lunivore.com

sebrose commented:
Thanks for writing this, Matt. 

As you know Gaspar and I went with Discovery, Formulation, and Automation. “Discovery” because “deliberate discovery, “Automation” because we have a constant battle with BDD being confused with testing. 

“Formulation” was harder. I explained how Gaspar convinced me in our book “Discovery”:

“When I first started working with Gáspár, I wondered why he kept using the word “formulate”. It’s not a term used by software professionals in general, or BDD practitioners in particular. He explained that it was the best word that he had found to describe the process of turning concrete examples into business-readable documentation. I was pretty sure there must be a better word, so I went away to try to think of one.
I tried substituting with “translate”, but there’s a lot of creative energy required to write a good BDD scenario. “Translate” didn’t do it justice, and nor did “convert”.
The Merriam_webster dictionary defines formulate as: to put into a systematized statement or expression. This is an accurate description of what we do when we convert examples into scenarios. This is an important task that is uniquely focused on by BDD practitioners and it deserves a precise name. Formulation is ideal!” 

The three practices we took directly from Liz’s work (“Step away from the tools” rather than “Three headed monster”). 

I’m with Tom with respect to the imperfection of naming, and with Liz that consensus may not be attainable. It feels like we’re getting to one of the later stages in Arlo’s naming process. 

The fourth practice feels useful. I need to think about how it fits into the overall picture.  “Habitation” certainly has a lot of good associations as a name. I don’t have Gojko’s book to hand, so will have to compare your words with his at a later date. 

Finally, I strongly agree with Richard’s observation about the iterative nature of the work. It would be good to make this clearer.

gojko commented:
how about 'documenting'? after a conversation/exploration, document the result.

also, instead of habitation, how about iterating or revising? that's kind of suggesting a recurring process of cleanup and revision?

mlvandijk commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,107 @@ menu: main
+* Working in rapid, small iterations to increase feedback and flow of value
+* Producing system documentation that is automatically validated against the system's behaviour
+
+We do this by focussing collaborative work around concrete, real-world, examples that illustrate how we want the system to behave. We use those examples to guide us from concept through to implementation.
```
```suggestion
We do this by focussing collaborative work around concrete, real-world examples that illustrate how we want the system to behave. We use those examples to guide us from concept through to implementation.
```

mlvandijk commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,107 @@ menu: main
+
+# Four practices
+
+Essentially, day-to-day BDD activity is a three-step, iterative, process:
```
```suggestion
Essentially, day-to-day BDD activity is a three-step, iterative process:
```

mlvandijk commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,107 @@ menu: main
+
+When we need to come back and maintain the system later, the automated examples will help us to understand what the system is currently doing, and to make changes safely without unintentionally breaking anything.
+
+This rapid, repeatable feedback reduces the burden of manual testing, freeing people up to do more interesting work, like exploratory testing.
```
```suggestion
This rapid, repeatable feedback reduces the burden of manual regression testing, freeing people up to do more interesting work, like exploratory testing.
```

mlvandijk replied:
Possibly link to a good explanation of exploratory testing somewhere?

mattwynne replied:
Yes, actually I was planning to add a terminology page for that. Where do you think would be a good place to point people towards?

mlvandijk commented on content/docs/bdd/myths.md:
```diff
@@ -0,0 +1,30 @@
+
+We see plenty of teams who try to leave the work of identifying examples and turning them into formulated scenarios to a single individual on the team.
+
+That's not BDD. Discovery work needs to be done collaboratively, bringing together representatives of the different specialists who all need to share understanding about what will be built.
```
Suggestion to put emphasis on *collaboratively*

mlvandijk commented on content/docs/bdd/overview.md:
```diff
@@ -1,90 +0,0 @@
-* Ops engineers
-* And probably some others
-
-The Cucumber way to do this is [Example Mapping](https://cucumber.io/blog/2015/12/08/example-mapping-introduction), a
```
Section on / link to example mapping would still be useful in the Discovery phase?

mattwynne replied:
Yes, I plan to make a whole new section on Discovery and the Example Mapping stuff will be in it.

mlvandijk commented on content/docs/bdd/overview.md:
```diff
@@ -1,90 +0,0 @@
-The examples they produce together can then become automated tests and living
-documentation of how the system behaves.
-
-# Test-Driven Development
```
Link between TDD / BDD seems useful to me

mlvandijk commented on content/docs/bdd/overview.md:
```diff
@@ -1,90 +0,0 @@
-Cucumber is the tool that can execute them.
-
-{{% note "Test after is not BDD" %}}
-Many people write tests after the code is written, even with Cucumber. This is
```
I like that you've put these in a myth buster page ;)

mattwynne replied:
:)

mlvandijk commented on content/docs/terms/_index.md:
```diff
@@ -0,0 +1,7 @@
+collapseChildren: true
+---
+
+There's loads of new terminology and jargon to learn as you learn about Cucumber and BDD. Here you'll find definitions of the common terms
```
```suggestion
There's loads of new terminology and jargon to learn as you learn about Cucumber and BDD. Here you'll find definitions of the common terms.
```

mlvandijk commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,107 @@ menu: main
 iconClass: fas fa-comments
```
What does this front-matter tag do?

mattwynne replied:
I think it adds an icon for this section on the main nav on the left?

mlvandijk commented on content/docs/terms/_index.md:
```diff
@@ -0,0 +1,7 @@
+---
+title: Terminology
+menu: main
+collapseChildren: true
```
What does this front-matter tag do?

mattwynne replied:
I have no idea - I was cargo-culting / copy-pasting from another main section.

aslakhellesoy replied:
It collapses the children in the left nav by default, rather than opening them.

mlvandijk commented:
@mattwynne Preview works as expected imho. You get a full preview of the docs and have to navigate to the relevant page(s) though. What, in your opinion, isn't working?

mlvandijk commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,107 @@ menu: main
+
+## BDD and agile
+
+We assume that your team are using some kind of agile methodology already, planning work in small increments of value like [User Stories]. BDD does not replace your existing agile process, it enhances it.
```
I like the links to terminology.

mattwynne replied:
:)

Yeah I'm hoping that can grow and become really useful.

mlvandijk reviewed (commented):
Suggestions on details, as I'll leave it to you all to describe BDD and it's terminology. Overall content is clear to me, but I'm not entirely new to BDD concepts anymore.

mattwynne committed:
grammar

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

mattwynne committed:
grammar

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

mattwynne committed:
Add Marit's feedback

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

mattwynne committed:
grammar

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

mattwynne commented:
> @mattwynne Preview works as expected imho. You get a full preview of the docs and have to navigate to the relevant page(s) though. What, in your opinion, isn't working?

@mlvandijk it does seem to be working again now, so I'm not sure. Previously, the link from the build check didn't seem to be showing the latest changes - it just showed the old, very short, index page for BDD from the master branch. Let's ignore it and hope it doesn't happen again :)

mattwynne committed:
Respond to feedback.

cbliard commented on content/docs/bdd/_index.md:
```diff
@@ -6,4 +6,102 @@ menu: main
+
+Over time, as your living documenation and codebase grow together, a symbiotic relationship develops between them. The documentation reflects the behaviour of the code, giving the whole team confidence about exactly what the system does. The code reflects the documentation, with rapid feedback for the developers if the code's behaviour strays outside the boundary defined by the living documentati ...
+
+This is where the investment of all that discovery, formulation and automation work pays off. The team are able to confidently and rapidly adapt the system --- and it's documentation --- to meet the business' evolving needs.
```
I think there is a typo here

```suggestion
This is where the investment of all that discovery, formulation and automation work pays off. The team are able to confidently and rapidly adapt the system --- and its documentation --- to meet the business' evolving needs.
```

mattwynne committed:
WIP - sharing latest changes with Seb

mattwynne committed:
Review with Seb

We're removing the fourth practice (notionally called "Habitation") for
now as we want more time to get it right, but the rest of it seems good
enough to ship.

We're also not delighted with the diagram, but it seems good enough for
now.

mattwynne commented:
Thanks for your feedback everyone! First iteration is live.

@gojko we've left the "habitation" practice / discipline out of this first iteration just because it felt less fully-baked than the rest of them, but we'd like to add it back in soon.

@richardlawrence thanks for the feedback about "continuous collaboration", this is indeed a common conceptual problem and needs to be adressed. @sebrose and I have worked on the copy to try and help with this - we added a section at the top about "rapid iterations" (trying to avoid using jargon like "single piece flow" and "pull-based") and tried to make the copy in the sections on the individual practices better reflect a just-in-time way of working. I can't help thinking a better diagram would help here, and I've been trying different options but so far I don't think I've quite nailed it.

aslakhellesoy commented:
Great work @mattwynne, @sebrose and everyone else who've helped out. Now that this PR is merged, where do you want to continue the conversation about how to improve this?

tastapod commented:
I have a couple of comments on this. I’ve been following the conversation but haven’t had the headspace to contribute in a meaningful way. I love the direction this is taking but there are some material changes I would propose. I’ll respond with a PR soon. Thanks everyone for some great content and discussion.

> On 24 Sep 2019, at 12:16, Aslak Hellesøy <notifications@github.com> wrote:
> 
> Great work @mattwynne <https://github.com/mattwynne>, @sebrose <https://github.com/sebrose> and everyone else who've helped out. Now that this PR is merged, where do you want to continue the conversation about how to improve this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cucumber/docs.cucumber.io/pull/376?email_source=notifications&email_token=AAAAY3GS2MEGAA25MIOALW3QLHZJLA5CNFSM4ITKXLMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7OABOQ#issuecomment-534511802>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAAAY3DFGAY6WGMJBI52CCTQLHZJLANCNFSM4ITKXLMA>.
>

gojko commented:
@tastapod you need a bigger head

tastapod commented:
That is the first time anyone has ever made that suggestion. Usually it’s the opposite.

> On 24 Sep 2019, at 12:32, Gojko Adzic <notifications@github.com> wrote:
> 
> @tastapod <https://github.com/tastapod> you need a bigger head
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cucumber/docs.cucumber.io/pull/376?email_source=notifications&email_token=AAAAY3H5W3LTUJ4KIRBMBBTQLH3DZA5CNFSM4ITKXLMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7OBG7I#issuecomment-534516605>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAAAY3HG6P2Y6M3UMOUAQ2TQLH3DZANCNFSM4ITKXLMA>.
>

mattwynne commented:
@aslakhellesoy I hope it's stable enough that we can talk about specific changes around PRs (looking forward to it @tastapod), but if there's more general feedback I don't mind people continuing the thread here. Sound good?

aslakhellesoy commented:
Ok @mattwynne here goes:

The [Discovery workshop](https://cucumber.io/docs/terms/discovery-workshop/) page is very light on details, and I'd expect a novice reader to have even more questions after reading this short page:

- How does this conversation happen?
- How often? When?
- Who should come?
- How long?

We have answers to all of those questions, but the page does not make them explicit:
- Talk about examples. Mention example mapping as _one_ way to do this. There is also event storming, impact mapping and [OOPSI](https://jennyjmar.com/2016/04/16/bdd-discovery-and-oopsi/)
- As late as possible before starting a new story
- 3 amigos
- Ca 30 min per story

I think we should either add more detail to this page, or simply delete it and point people to the Example Mapping page instead.

mattwynne commented:
Thanks @aslakhellesoy. My plan with the new /docs/terms section is to mostly use it as a way to let people navigate around the content rather than *being* the content, so I was thinking those pages would be quite short, with signposts to where people can find out more. Hugo seems good at this, but we need the content first!

So I'd expect the gaps you've identified there to be filled in when we build out the /docs/bdd/discovery section. My plan is to write out sections for each of the three practices over the coming mnonths, and clean up all the existing content we have here that overlaps with them.

mattwynne commented:
@tastapod we're going to be committing the points in this article to video in the next few months, so I'd love to hear your feedback even in rough form before we do that.

What about we just speak on the phone?
